File: Libreria.java
import java.util.ArrayList;
import java.util.List;

public class Libreria {
    private List<Usuario> usuarios;
    private List<Libro> libros;

    public Libreria() {
        this.usuarios = new ArrayList<>();
        this.libros = new ArrayList<>();
    }

    public void registrarUsuario(String nombre) {
        Usuario usuario = new Usuario(nombre);
        usuarios.add(usuario);
    }

    public void registrarLibro(String nombre, String autor) {
        Libro libro = new Libro(nombre, autor);
        libros.add(libro);
    }

    public void rentarLibro(int idUsuario, int idLibro) {
        Usuario usuario = buscarUsuarioPorId(idUsuario);
        Libro libro = buscarLibroPorId(idLibro);

        if (usuario != null && libro != null && !libro.isRentado()) {
            usuario.rentarLibro(libro);
            libro.setRentado(true);
            System.out.println("El libro se rento con exito.");
        } else {
            System.out.println("No se pudo rentar el libro.");
        }
    }

    public void listarUsuarios() {
        System.out.println("*** USUARIOS REGISTRADOS ***");
        for (Usuario usuario : usuarios) {
            System.out.println(usuario.getNombre());
        }
    }

    public void listarLibros() {
        System.out.println("*** LIBROS REGISTRADOS ***");
        for (Libro libro : libros) {
            System.out.println(libro.getNombre() + " - " + libro.getAutor());
        }
    }

    public void listarUsuariosConLibros() {
        System.out.println("*** USUARIOS CON LIBROS RENTADOS ***");
        for (Usuario usuario : usuarios) {
            if (!usuario.getLibrosRentados().isEmpty()) {
                System.out.println(usuario.getNombre());
            }
        }
    }

    public void listarLibrosNoRentados() {
        System.out.println("*** LIBROS NO RENTADOS ***");
        for (Libro libro : libros) {
            if (!libro.isRentado()) {
                System.out.println(libro.getNombre() + " - " + libro.getAutor());
            }
        }
    }

    public void listarLibrosRentados() {
        System.out.println("*** LIBROS RENTADOS ***");
        for (Libro libro : libros) {
            if (libro.isRentado()) {
                System.out.println(libro.getNombre() + " - " + libro.getAutor());
            }
        }
    }

    private Usuario buscarUsuarioPorId(int id) {
        for (Usuario usuario : usuarios) {
            if (usuario.getId() == id) {
                return usuario;
            }
        }
        return null;
    }

    private Libro buscarLibroPorId(int id) {
        for (Libro libro : libros) {
            if (libro.getId() == id) {
                return libro;
            }
        }
        return null;
    }
}

File: Libreria.py
class Usuario:
    def __init__(self, nombre):
        self.id = id(self)
        self.nombre = nombre
        self.libros_rentados = []

    def rentar_libro(self, libro):
        self.libros_rentados.append(libro)

    def __str__(self):
        return self.nombre


class Libro:
    def __init__(self, nombre, autor):
        self.id = id(self)
        self.nombre = nombre
        self.autor = autor
        self.rentado = False

    def set_rentado(self, rentado):
        self.rentado = rentado

    def __str__(self):
        return f"{self.nombre} - {self.autor}"


class Libreria:
    def __init__(self):
        self.usuarios = []
        self.libros = []

    def registrar_usuario(self, nombre):
        usuario = Usuario(nombre)
        self.usuarios.append(usuario)

    def registrar_libro(self, nombre, autor):
        libro = Libro(nombre, autor)
        self.libros.append(libro)

    def rentar_libro(self, id_usuario, id_libro):
        usuario = self.buscar_usuario_por_id(id_usuario)
        libro = self.buscar_libro_por_id(id_libro)

        if usuario and libro and not libro.rentado:
            usuario.rentar_libro(libro)
            libro.set_rentado(True)
            print("El libro se rento con exito.")
        else:
            print("No se pudo rentar el libro.")

    def listar_usuarios(self):
        print("*** USUARIOS REGISTRADOS ***")
        for usuario in self.usuarios:
            print(usuario)

    def listar_libros(self):
        print("*** LIBROS REGISTRADOS ***")
        for libro in self.libros:
            print(libro)

    def listar_usuarios_con_libros(self):
        print("*** USUARIOS CON LIBROS RENTADOS ***")
        for usuario in self.usuarios:
            if usuario.libros_rentados:
                print(usuario)

    def listar_libros_no_rentados(self):
        print("*** LIBROS NO RENTADOS ***")
        for libro in self.libros:
            if not libro.rentado:
                print(libro)

    def listar_libros_rentados(self):
        print("*** LIBROS RENTADOS ***")
        for libro in self.libros:
            if libro.rentado:
                print(libro)

    def buscar_usuario_por_id(self, id):
        for usuario in self.usuarios:
            if usuario.id == id:
                return usuario
        return None

    def buscar_libro_por_id(self, id):
        for libro in self.libros:
            if libro.id == id:
                return libro
        return None